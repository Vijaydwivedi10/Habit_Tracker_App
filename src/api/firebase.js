import firebase from 'firebase/app';

import 'firebase/auth';
import 'firebase/database';
import 'firebase/firestore';

const firebaseConfig = {
  apiKey: "AIzaSyDi_KD92NN_-SIKHq9viP-RKofauNFP5b0",
  authDomain: "habit-tracker-409bd.firebaseapp.com",
  databaseURL: "https://habit-tracker-409bd-default-rtdb.asia-southeast1.firebasedatabase.app",
  projectId: "habit-tracker-409bd",
  storageBucket: "habit-tracker-409bd.appspot.com",
  messagingSenderId: "146003024957",
  appId: "1:146003024957:web:34546d376a5c7bd71bde54",
  measurementId: "G-Z67XDDMWMB"
};


if (!firebase.apps.length) {
  firebase.initializeApp(firebaseConfig);
} else {
  firebase.app();
}

const app = {
  firebase,
  auth: firebase.auth(),
  firestore: firebase.firestore(),
  db: firebase.database(),
};

export default app;
