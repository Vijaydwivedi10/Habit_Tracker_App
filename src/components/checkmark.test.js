import * as React from 'react';
import { render, fireEvent,waitFor } from '@testing-library/react';
import { IconButton } from '@material-ui/core';
import {
  CheckBox as CompletedCheckmarkIcon,
  CheckBoxOutlineBlank as EmptyCheckmarkIcon,
  IndeterminateCheckBox as FailedCheckmarkIcon,
} from '@material-ui/icons';
import { useUpdateCheckmarkInDbMutate } from 'api/checkmarks';
import { COMPLETED, EMPTY, FAILED } from 'data/constants';
import { debounce } from 'lodash';

const variants = {
  completed: {
    icon: <CompletedCheckmarkIcon />,
    label: 'completed',
    color: 'primary',
  },
  failed: {
    icon: <FailedCheckmarkIcon />,
    label: 'failed',
    color: 'secondary',
  },
  empty: {
    icon: <EmptyCheckmarkIcon />,
    label: 'empty',
    color: 'default',
  },
};

function Checkmark({ id, initialValue, habitId, date, disabled }) {
  const [value, setValue] = React.useState(initialValue);

  React.useEffect(() => {
    setValue(initialValue);
  }, [initialValue]);

  const updateCheckmarkInDbMutate = useUpdateCheckmarkInDbMutate();

  // Debounced update function
  const debouncedUpdate = React.useRef(
    debounce(({ id, newValue }) => {
      updateCheckmarkInDbMutate({
        checkmarkId: id,
        value: newValue,
        habitId,
        date,
      });
    }, 200)
  ).current;

  // Handles clicking on checkmark
  const handleClick = () => {
    const newValue = getNewValue(value);

    // Update the value locally, so that the icon changes
    setValue(newValue);
    // Update is debounced so when user is clicking very fast on the checkmark
    // only the last call will be invoked to hit the database.
    debouncedUpdate({ id, newValue });
  };

  const { icon, label, color } = variants[value];

  return (
    <IconButton
      aria-label={label}
      color={color}
      onClick={handleClick}
      disabled={disabled}
    >
      {icon}
    </IconButton>
  );
}

function getNewValue(currentValue) {
  const values = [COMPLETED, FAILED, EMPTY];

  return values[(values.indexOf(currentValue) + 1) % values.length];
}

describe('Checkmark', () => {
  const mockUpdateCheckmarkInDbMutate = jest.fn();

  beforeEach(() => {
    mockUpdateCheckmarkInDbMutate.mockClear();
    jest.spyOn(require('api/checkmarks'), 'useUpdateCheckmarkInDbMutate').mockReturnValue(mockUpdateCheckmarkInDbMutate);
  });

  it('renders the completed checkmark correctly', () => {
    const { container } = render(<Checkmark id="checkmark-id" initialValue="completed" habitId="habit-id" date="2022-01-01" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the failed checkmark correctly', () => {
    const { container } = render(<Checkmark id="checkmark-id" initialValue="failed" habitId="habit-id" date="2022-01-01" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the empty checkmark correctly', () => {
    const { container } = render(<Checkmark id="checkmark-id" initialValue="empty" habitId="habit-id" date="2022-01-01" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('updates the value when clicked', () => {
    const { getByRole } = render(
    <Checkmark id="checkmark-id" initialValue="completed" habitId="habit-id" date="2022-01-01" />
    );
    fireEvent.click(getByRole('button'));
    expect(getByRole('button')).toHaveAttribute('aria-label', 'failed');
    });
    
    it('updates the value with correct debounce', async () => {
    const updateCheckmarkInDbMutateMock = jest.fn();
    const { getByRole } = render(
    <Checkmark
         id="checkmark-id"
         initialValue="completed"
         habitId="habit-id"
         date="2022-01-01"
       />
    );
    fireEvent.click(getByRole('button'));
    fireEvent.click(getByRole('button'));
    fireEvent.click(getByRole('button'));
    await waitFor(() => {
    expect(updateCheckmarkInDbMutateMock).toHaveBeenCalledTimes(0);
    
    });
    });
    
    it('disables the button when disabled prop is true', () => {
    const { getByRole } = render(
    <Checkmark id="checkmark-id" initialValue="completed" habitId="habit-id" date="2022-01-01" disabled />
    );
    expect(getByRole('button')).toBeDisabled();
    });
    
    it('does not update the value when disabled', () => {
    const updateCheckmarkInDbMutateMock = jest.fn();
    const { getByRole } = render(
    <Checkmark id="checkmark-id" initialValue="completed" habitId="habit-id" date="2022-01-01" disabled />
    );
    fireEvent.click(getByRole('button'));
    expect(getByRole('button')).toHaveAttribute('aria-label', 'completed');
    expect(updateCheckmarkInDbMutateMock).not.toHaveBeenCalled();
    });
    });
    
    describe('getNewValue', () => {
    it('returns the next value in the cycle', () => {
    expect(getNewValue('completed')).toBe('failed');
    expect(getNewValue('failed')).toBe('empty');
    expect(getNewValue('empty')).toBe('completed');
    });
    });