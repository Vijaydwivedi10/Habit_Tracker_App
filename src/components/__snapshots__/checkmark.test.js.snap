// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkmark renders the completed checkmark correctly 1`] = `
<button
  aria-label="completed"
  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      data-testid="CheckBoxIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Checkmark renders the empty checkmark correctly 1`] = `
<button
  aria-label="empty"
  class="MuiButtonBase-root MuiIconButton-root"
  tabindex="0"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      data-testid="CheckBoxOutlineBlankIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`Checkmark renders the failed checkmark correctly 1`] = `
<button
  aria-label="failed"
  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      data-testid="IndeterminateCheckBoxIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;
